// @herb-tools/linter/dist/src/linter.js
import { Location } from "@herb-tools/core";
import { IdentityPrinter } from "@herb-tools/printer";
import { minimatch } from "minimatch";
import { rules } from "./rules.js";
import { findNodeByLocation } from "./rules/rule-utils.js";
import { parseHerbDisableLine } from "./herb-disable-comment-utils.js";
import { hasLinterIgnoreDirective } from "./linter-ignore.js";
import { ParserNoErrorsRule } from "./rules/parser-no-errors.js";
import { DEFAULT_RULE_CONFIG } from "./types.js";
export class Linter {
    rules;
    allAvailableRules;
    herb;
    offenses;
    config;
    /**
     * Creates a new Linter instance with automatic rule filtering based on config.
     *
     * @param herb - The Herb backend instance for parsing and lexing
     * @param config - Optional full Config instance for rule filtering, severity overrides, and path-based filtering
     * @param customRules - Optional array of custom rules to include alongside built-in rules
     * @returns A configured Linter instance
     */
    static from(herb, config, customRules) {
        const allRules = customRules ? [...rules, ...customRules] : rules;
        const filteredRules = config?.linter?.rules
            ? Linter.filterRulesByConfig(allRules, config.linter.rules)
            : undefined;
        return new Linter(herb, filteredRules, config, allRules);
    }
    /**
     * Creates a new Linter instance.
     *
     * For most use cases, prefer `Linter.from()` which handles config-based filtering.
     * Use this constructor directly when you need explicit control over rules.
     *
     * @param herb - The Herb backend instance for parsing and lexing
     * @param rules - Array of rule classes (Parser/AST or Lexer) to use. If not provided, uses default rules.
     * @param config - Optional full Config instance for severity overrides and path-based rule filtering
     * @param allAvailableRules - Optional array of ALL available rules (including disabled) for herb:disable validation
     */
    constructor(herb, rules, config, allAvailableRules) {
        this.herb = herb;
        this.config = config;
        this.rules = rules !== undefined ? rules : this.getDefaultRules();
        this.allAvailableRules = allAvailableRules !== undefined ? allAvailableRules : this.rules;
        this.offenses = [];
    }
    /**
     * Filters rules based on default config and optional user config overrides.
     *
     * Priority:
     * 1. User config override (if rule config exists in userRulesConfig)
     * 2. Default config from rule's defaultConfig getter
     *
     * @param allRules - All available rule classes to filter from
     * @param userRulesConfig - Optional user configuration for rules
     * @returns Filtered array of rule classes that should be enabled
     */
    static filterRulesByConfig(allRules, userRulesConfig) {
        return allRules.filter(ruleClass => {
            const instance = new ruleClass();
            const ruleName = instance.name;
            const defaultEnabled = instance.defaultConfig?.enabled ?? DEFAULT_RULE_CONFIG.enabled;
            const userRuleConfig = userRulesConfig?.[ruleName];
            if (userRuleConfig !== undefined) {
                return userRuleConfig.enabled !== false;
            }
            return defaultEnabled;
        });
    }
    /**
     * Returns the default set of rule classes used by the linter.
     * These are the rules enabled when no custom rules are provided.
     * Filters all available rules to only include those enabled by default.
     * @returns Array of default rule classes
     */
    getDefaultRules() {
        return Linter.filterRulesByConfig(rules);
    }
    /**
     * Returns all available rule classes that can be referenced in herb:disable comments.
     * This includes all rules that exist, regardless of whether they're currently enabled.
     * Includes both built-in rules and any loaded custom rules.
     * @returns Array of all available rule classes
     */
    getAvailableRules() {
        return this.allAvailableRules;
    }
    /**
     * Meta-linting rules for herb:disable comments cannot be disabled
     * This ensures that invalid herb:disable comments are always caught
     */
    get nonExcludableRules() {
        return [
            "herb-disable-comment-valid-rule-name",
            "herb-disable-comment-no-redundant-all",
            "herb-disable-comment-no-duplicate-rules",
            "herb-disable-comment-malformed",
            "herb-disable-comment-missing-rules",
            "herb-disable-comment-unnecessary"
        ];
    }
    getRuleCount() {
        return this.rules.length;
    }
    /**
     * Type guard to check if a rule is a LexerRule
     */
    isLexerRule(rule) {
        return rule.constructor.type === "lexer";
    }
    /**
     * Type guard to check if a rule is a SourceRule
     */
    isSourceRule(rule) {
        return rule.constructor.type === "source";
    }
    /**
     * Execute a single rule and return its unbound offenses.
     * Handles rule type checking (Lexer/Parser/Source) and isEnabled checks.
     */
    executeRule(rule, parseResult, lexResult, source, context) {
        if (this.config && context?.fileName) {
            if (!this.config.isRuleEnabledForPath(rule.name, context.fileName)) {
                return [];
            }
        }
        if (context?.fileName && !this.config?.linter?.rules?.[rule.name]?.exclude) {
            const defaultExclude = rule.defaultConfig?.exclude ?? DEFAULT_RULE_CONFIG.exclude;
            if (defaultExclude && defaultExclude.length > 0) {
                const isExcluded = defaultExclude.some((pattern) => minimatch(context.fileName, pattern));
                if (isExcluded) {
                    return [];
                }
            }
        }
        let isEnabled = true;
        let ruleOffenses;
        if (this.isLexerRule(rule)) {
            if (rule.isEnabled) {
                isEnabled = rule.isEnabled(lexResult, context);
            }
            if (isEnabled) {
                ruleOffenses = rule.check(lexResult, context);
            }
            else {
                ruleOffenses = [];
            }
        }
        else if (this.isSourceRule(rule)) {
            if (rule.isEnabled) {
                isEnabled = rule.isEnabled(source, context);
            }
            if (isEnabled) {
                ruleOffenses = rule.check(source, context);
            }
            else {
                ruleOffenses = [];
            }
        }
        else {
            if (rule.isEnabled) {
                isEnabled = rule.isEnabled(parseResult, context);
            }
            if (isEnabled) {
                ruleOffenses = rule.check(parseResult, context);
            }
            else {
                ruleOffenses = [];
            }
        }
        return ruleOffenses;
    }
    filterOffenses(ruleOffenses, ruleName, ignoredOffensesByLine, herbDisableCache, ignoreDisableComments) {
        const kept = [];
        const ignored = [];
        const wouldBeIgnored = [];
        if (this.nonExcludableRules.includes(ruleName)) {
            return { kept: ruleOffenses, ignored: [], wouldBeIgnored: [] };
        }
        if (ignoreDisableComments) {
            for (const offense of ruleOffenses) {
                const line = offense.location.start.line;
                const disabledRules = herbDisableCache?.get(line) || [];
                if (disabledRules.includes(ruleName) || disabledRules.includes("all")) {
                    wouldBeIgnored.push(offense);
                }
            }
            return { kept: ruleOffenses, ignored: [], wouldBeIgnored };
        }
        for (const offense of ruleOffenses) {
            const line = offense.location.start.line;
            const disabledRules = herbDisableCache?.get(line) || [];
            if (disabledRules.includes(ruleName) || disabledRules.includes("all")) {
                ignored.push(offense);
                if (ignoredOffensesByLine) {
                    if (!ignoredOffensesByLine.has(line)) {
                        ignoredOffensesByLine.set(line, new Set());
                    }
                    const usedRuleName = disabledRules.includes(ruleName) ? ruleName : "all";
                    ignoredOffensesByLine.get(line).add(usedRuleName);
                }
                continue;
            }
            kept.push(offense);
        }
        return { kept, ignored, wouldBeIgnored: [] };
    }
    /**
     * Lint source code using Parser/AST, Lexer, and Source rules.
     * @param source - The source code to lint
     * @param context - Optional context for linting (e.g., fileName for distinguishing files vs snippets)
     */
    lint(source, context) {
        this.offenses = [];
        let ignoredCount = 0;
        let wouldBeIgnoredCount = 0;
        const parseResult = this.herb.parse(source, { track_whitespace: true });
        // Check for file-level ignore directive using visitor
        if (hasLinterIgnoreDirective(parseResult)) {
            return {
                offenses: [],
                errors: 0,
                warnings: 0,
                info: 0,
                hints: 0,
                ignored: 0
            };
        }
        const lexResult = this.herb.lex(source);
        const hasParserErrors = parseResult.recursiveErrors().length > 0;
        const sourceLines = source.split("\n");
        const ignoredOffensesByLine = new Map();
        const herbDisableCache = new Map();
        if (hasParserErrors) {
            const hasParserRule = this.rules.find(RuleClass => (new RuleClass()).name === "parser-no-errors");
            if (hasParserRule) {
                const rule = new ParserNoErrorsRule();
                const offenses = rule.check(parseResult);
                this.offenses.push(...offenses);
            }
            return {
                offenses: this.offenses,
                errors: this.offenses.filter(o => o.severity === "error").length,
                warnings: this.offenses.filter(o => o.severity === "warning").length,
                info: this.offenses.filter(o => o.severity === "info").length,
                hints: this.offenses.filter(o => o.severity === "hint").length,
                ignored: 0
            };
        }
        for (let i = 0; i < sourceLines.length; i++) {
            const line = sourceLines[i];
            if (line.includes("herb:disable")) {
                const herbDisable = parseHerbDisableLine(line);
                herbDisableCache.set(i + 1, herbDisable?.ruleNames || []);
            }
        }
        context = {
            ...context,
            validRuleNames: this.getAvailableRules().map(RuleClass => new RuleClass().name),
            ignoredOffensesByLine
        };
        const regularRules = this.rules.filter(RuleClass => {
            const rule = new RuleClass();
            return rule.name !== "herb-disable-comment-unnecessary";
        });
        for (const RuleClass of regularRules) {
            const rule = new RuleClass();
            const unboundOffenses = this.executeRule(rule, parseResult, lexResult, source, context);
            const boundOffenses = this.bindSeverity(unboundOffenses, rule.name);
            const { kept, ignored, wouldBeIgnored } = this.filterOffenses(boundOffenses, rule.name, ignoredOffensesByLine, herbDisableCache, context?.ignoreDisableComments);
            ignoredCount += ignored.length;
            wouldBeIgnoredCount += wouldBeIgnored.length;
            this.offenses.push(...kept);
        }
        const unnecessaryRuleClass = this.rules.find(RuleClass => {
            const rule = new RuleClass();
            return rule.name === "herb-disable-comment-unnecessary";
        });
        if (unnecessaryRuleClass) {
            const unnecessaryRule = new unnecessaryRuleClass();
            const unboundOffenses = unnecessaryRule.check(parseResult, context);
            const boundOffenses = this.bindSeverity(unboundOffenses, unnecessaryRule.name);
            this.offenses.push(...boundOffenses);
        }
        const finalOffenses = this.offenses;
        const errors = finalOffenses.filter(offense => offense.severity === "error").length;
        const warnings = finalOffenses.filter(offense => offense.severity === "warning").length;
        const info = finalOffenses.filter(offense => offense.severity === "info").length;
        const hints = finalOffenses.filter(offense => offense.severity === "hint").length;
        const result = {
            offenses: finalOffenses,
            errors,
            warnings,
            info,
            hints,
            ignored: ignoredCount
        };
        if (wouldBeIgnoredCount > 0) {
            result.wouldBeIgnored = wouldBeIgnoredCount;
        }
        return result;
    }
    /**
     * Bind severity to unbound offenses based on rule's defaultConfig and user config overrides.
     *
     * Priority:
     * 1. User config severity override (if specified in config)
     * 2. Rule's default severity (from defaultConfig.severity)
     *
     * @param unboundOffenses - Array of offenses without severity
     * @param ruleName - Name of the rule that produced the offenses
     * @returns Array of offenses with severity bound
     */
    bindSeverity(unboundOffenses, ruleName) {
        const RuleClass = this.rules.find(rule => {
            const instance = new rule();
            return instance.name === ruleName;
        });
        if (!RuleClass) {
            return unboundOffenses.map(offense => ({
                ...offense,
                severity: "error"
            }));
        }
        const ruleInstance = new RuleClass();
        const defaultSeverity = ruleInstance.defaultConfig?.severity ?? DEFAULT_RULE_CONFIG.severity;
        const userRuleConfig = this.config?.linter?.rules?.[ruleName];
        const severity = userRuleConfig?.severity ?? defaultSeverity;
        return unboundOffenses.map(offense => ({
            ...offense,
            severity
        }));
    }
    /**
     * Automatically fix offenses in the source code.
     * Uses AST mutation for parser rules and token mutation for lexer rules.
     * @param source - The source code to fix
     * @param context - Optional context for linting (e.g., fileName)
     * @param offensesToFix - Optional array of specific offenses to fix. If not provided, all fixable offenses will be fixed.
     * @returns AutofixResult containing the corrected source and lists of fixed/unfixed offenses
     */
    autofix(source, context, offensesToFix) {
        const lintResult = offensesToFix ? { offenses: offensesToFix } : this.lint(source, context);
        const parserOffenses = [];
        const lexerOffenses = [];
        const sourceOffenses = [];
        for (const offense of lintResult.offenses) {
            const RuleClass = this.rules.find(rule => {
                const instance = new rule();
                return instance.name === offense.rule;
            });
            if (!RuleClass)
                continue;
            if (RuleClass.type === "lexer") {
                lexerOffenses.push(offense);
            }
            else if (RuleClass.type === "source") {
                sourceOffenses.push(offense);
            }
            else {
                parserOffenses.push(offense);
            }
        }
        let currentSource = source;
        const fixed = [];
        const unfixed = [];
        if (parserOffenses.length > 0) {
            const parseResult = this.herb.parse(currentSource, { track_whitespace: true });
            for (const offense of parserOffenses) {
                const RuleClass = this.rules.find(rule => new rule().name === offense.rule);
                if (!RuleClass) {
                    unfixed.push(offense);
                    continue;
                }
                const rule = new RuleClass();
                if (!rule.autofix) {
                    unfixed.push(offense);
                    continue;
                }
                if (offense.autofixContext) {
                    const originalNodeType = offense.autofixContext.node.type;
                    const location = offense.autofixContext.node.location ? Location.from(offense.autofixContext.node.location) : offense.location;
                    const freshNode = findNodeByLocation(parseResult.value, location, (node) => node.type === originalNodeType);
                    if (freshNode) {
                        offense.autofixContext.node = freshNode;
                    }
                    else {
                        unfixed.push(offense);
                        continue;
                    }
                }
                const fixedResult = rule.autofix(offense, parseResult, context);
                if (fixedResult) {
                    fixed.push(offense);
                }
                else {
                    unfixed.push(offense);
                }
            }
            if (fixed.length > 0) {
                const printer = new IdentityPrinter();
                currentSource = printer.print(parseResult.value);
            }
        }
        if (sourceOffenses.length > 0) {
            const sortedSourceOffenses = sourceOffenses.sort((a, b) => {
                if (a.location.start.line !== b.location.start.line) {
                    return b.location.start.line - a.location.start.line;
                }
                return b.location.start.column - a.location.start.column;
            });
            for (const offense of sortedSourceOffenses) {
                const RuleClass = this.rules.find(rule => new rule().name === offense.rule);
                if (!RuleClass) {
                    unfixed.push(offense);
                    continue;
                }
                const rule = new RuleClass();
                if (!rule.autofix) {
                    unfixed.push(offense);
                    continue;
                }
                const correctedSource = rule.autofix(offense, currentSource, context);
                if (correctedSource) {
                    currentSource = correctedSource;
                    fixed.push(offense);
                }
                else {
                    unfixed.push(offense);
                }
            }
        }
        return {
            source: currentSource,
            fixed,
            unfixed
        };
    }
}
