// @herb-tools/linter/dist/src/types.js
/**
 * Default configuration for rules when defaultConfig is not specified.
 * Custom rules can omit defaultConfig and will use these defaults.
 */
export const DEFAULT_RULE_CONFIG = {
    enabled: true,
    severity: "error",
    exclude: []
};
/**
 * Base class for parser rules.
 */
export class ParserRule {
    static type = "parser";
    /** Indicates whether this rule supports autofix. Defaults to false. */
    static autocorrectable = false;
    get defaultConfig() {
        return DEFAULT_RULE_CONFIG;
    }
}
/**
 * Base class for lexer rules.
 */
export class LexerRule {
    static type = "lexer";
    /** Indicates whether this rule supports autofix. Defaults to false. */
    static autocorrectable = false;
    get defaultConfig() {
        return DEFAULT_RULE_CONFIG;
    }
}
/**
 * Default context object with all keys defined but set to undefined
 */
export const DEFAULT_LINT_CONTEXT = {
    fileName: undefined,
    validRuleNames: undefined,
    ignoredOffensesByLine: undefined,
    ignoreDisableComments: undefined
};
export class SourceRule {
    static type = "source";
    /** Indicates whether this rule supports autofix. Defaults to false. */
    static autocorrectable = false;
    get defaultConfig() {
        return DEFAULT_RULE_CONFIG;
    }
}
