// @herb-tools/linter/dist/src/rules/erb-require-trailing-newline.js
import { SourceRule } from "../types.js";
import { BaseSourceRuleVisitor, createEndOfFileLocation } from "./rule-utils.js";
class ERBRequireTrailingNewlineVisitor extends BaseSourceRuleVisitor {
    visitSource(source) {
        if (source.length === 0)
            return;
        if (!this.context.fileName)
            return;
        if (!source.endsWith('\n')) {
            this.addOffense("File must end with trailing newline.", createEndOfFileLocation(source));
        }
        else if (source.endsWith('\n\n')) {
            this.addOffense("File must end with exactly one trailing newline.", createEndOfFileLocation(source));
        }
    }
}
export class ERBRequireTrailingNewlineRule extends SourceRule {
    static autocorrectable = true;
    name = "erb-require-trailing-newline";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error"
        };
    }
    check(source, context) {
        const visitor = new ERBRequireTrailingNewlineVisitor(this.name, context);
        visitor.visit(source);
        return visitor.offenses;
    }
    autofix(_offense, source, _context) {
        return source.trimEnd() + "\n";
    }
}
