// @herb-tools/linter/dist/src/rules/erb-right-trim.js
import { BaseRuleVisitor } from "./rule-utils.js";
import { ParserRule } from "../types.js";
class ERBRightTrimVisitor extends BaseRuleVisitor {
    visitERBNode(node) {
        if (!node.tag_closing)
            return;
        const trimClosing = node.tag_closing.value;
        if (trimClosing !== "=%>")
            return;
        this.addOffense("Use `-%>` instead of `=%>` for right-trimming. The `=%>` syntax is obscure and not well-supported in most ERB engines.", node.tag_closing.location, { node });
    }
}
export class ERBRightTrimRule extends ParserRule {
    static autocorrectable = true;
    name = "erb-right-trim";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error"
        };
    }
    check(result, context) {
        const visitor = new ERBRightTrimVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
    autofix(offense, result, _context) {
        if (!offense.autofixContext)
            return null;
        const { node } = offense.autofixContext;
        if (!node.tag_closing)
            return null;
        const closing = node.tag_closing;
        if (closing.value === "=%>") {
            closing.value = "-%>";
            return result;
        }
        if (closing.value === "-%>") {
            closing.value = "%>";
            return result;
        }
        return null;
    }
}
