// @herb-tools/linter/dist/src/rules/herb-disable-comment-valid-rule-name.js
import { ParserRule } from "../types.js";
import { HerbDisableCommentParsedVisitor } from "./herb-disable-comment-base.js";
import { didyoumean } from "@herb-tools/core";
class HerbDisableCommentValidRuleNameVisitor extends HerbDisableCommentParsedVisitor {
    validRuleNames = new Set();
    validRuleNamesList = [];
    constructor(ruleName, validRuleNames, context) {
        super(ruleName, context);
        this.validRuleNames = new Set([...validRuleNames, "all"]);
        this.validRuleNamesList = Array.from(this.validRuleNames);
    }
    checkParsedHerbDisable(node, _content, herbDisable) {
        herbDisable.ruleNameDetails.forEach(ruleDetail => {
            if (this.validRuleNames.has(ruleDetail.name))
                return;
            const suggestion = didyoumean(ruleDetail.name, this.validRuleNamesList);
            const message = suggestion
                ? `Unknown rule \`${ruleDetail.name}\`. Did you mean \`${suggestion}\`?`
                : `Unknown rule \`${ruleDetail.name}\`.`;
            const location = this.createRuleNameLocation(node, ruleDetail);
            this.addOffenseWithFallback(message, location, node);
        });
    }
}
export class HerbDisableCommentValidRuleNameRule extends ParserRule {
    name = "herb-disable-comment-valid-rule-name";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "warning"
        };
    }
    check(result, context) {
        const validRuleNames = context?.validRuleNames;
        if (!validRuleNames)
            return [];
        if (validRuleNames.length === 0)
            return [];
        const visitor = new HerbDisableCommentValidRuleNameVisitor(this.name, validRuleNames, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
