// @herb-tools/linter/dist/src/rules/herb-disable-comment-missing-rules.js
import { ParserRule } from "../types.js";
import { HerbDisableCommentBaseVisitor } from "./herb-disable-comment-base.js";
import { parseHerbDisableContent } from "../herb-disable-comment-utils.js";
class HerbDisableCommentMissingRulesVisitor extends HerbDisableCommentBaseVisitor {
    checkHerbDisableComment(node, content) {
        const herbDisable = parseHerbDisableContent(content);
        if (herbDisable)
            return;
        const emptyFormat = /^\s*herb:disable\s*$/;
        if (!emptyFormat.test(content))
            return;
        this.addOffense(`\`herb:disable\` comment is missing rule names. Specify \`all\` or list specific rules to disable.`, node.location);
    }
}
export class HerbDisableCommentMissingRulesRule extends ParserRule {
    name = "herb-disable-comment-missing-rules";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error"
        };
    }
    check(result, context) {
        const visitor = new HerbDisableCommentMissingRulesVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
