// @herb-tools/linter/dist/src/rules/herb-disable-comment-no-duplicate-rules.js
import { ParserRule } from "../types.js";
import { HerbDisableCommentParsedVisitor } from "./herb-disable-comment-base.js";
class HerbDisableCommentNoDuplicateRulesVisitor extends HerbDisableCommentParsedVisitor {
    checkParsedHerbDisable(node, _content, herbDisable) {
        const seenRules = new Map();
        herbDisable.ruleNameDetails.forEach((ruleDetail, index) => {
            const firstIndex = seenRules.get(ruleDetail.name);
            if (firstIndex !== undefined) {
                const location = this.createRuleNameLocation(node, ruleDetail);
                const message = `Duplicate rule \`${ruleDetail.name}\` in \`herb:disable\` comment. Remove the duplicate.`;
                this.addOffenseWithFallback(message, location, node);
                return;
            }
            seenRules.set(ruleDetail.name, index);
        });
    }
}
export class HerbDisableCommentNoDuplicateRulesRule extends ParserRule {
    name = "herb-disable-comment-no-duplicate-rules";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "warning"
        };
    }
    check(result, context) {
        const visitor = new HerbDisableCommentNoDuplicateRulesVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
