// @herb-tools/linter/dist/src/rules/herb-disable-comment-base.js
import { BaseRuleVisitor } from "./rule-utils.js";
import { Location } from "@herb-tools/core";
import { parseHerbDisableContent } from "../herb-disable-comment-utils.js";
/**
 * Base visitor class for herb:disable comment validation rules.
 * Handles common patterns like checking ERB comments and parsing herb:disable content.
 */
export class HerbDisableCommentBaseVisitor extends BaseRuleVisitor {
    constructor(ruleName, context) {
        super(ruleName, context);
    }
    visitERBContentNode(node) {
        if (node.tag_opening?.value !== "<%#")
            return;
        const content = node.content?.value;
        if (!content)
            return;
        this.checkHerbDisableComment(node, content);
    }
    /**
     * Helper to create a precise location for a specific rule name within the comment.
     * Returns null if content location is not available.
     */
    createRuleNameLocation(node, ruleDetail) {
        const contentLocation = node.content?.location;
        if (!contentLocation)
            return null;
        const startLine = contentLocation.start.line;
        const startColumn = contentLocation.start.column + ruleDetail.offset;
        return Location.from(startLine, startColumn, startLine, startColumn + ruleDetail.length);
    }
    /**
     * Helper to add an offense with a fallback to node location if precise location unavailable.
     */
    addOffenseWithFallback(message, preciseLocation, node) {
        this.addOffense(message, preciseLocation || node.location);
    }
}
/**
 * Base visitor for rules that need to process parsed herb:disable comments.
 * Only calls the abstract method if the content successfully parses as a herb:disable comment.
 */
export class HerbDisableCommentParsedVisitor extends HerbDisableCommentBaseVisitor {
    checkHerbDisableComment(node, content) {
        const herbDisable = parseHerbDisableContent(content);
        if (!herbDisable)
            return;
        this.checkParsedHerbDisable(node, content, herbDisable);
    }
}
