// @herb-tools/linter/dist/src/rules/html-attribute-values-require-quotes.js
import { Token, Location } from "@herb-tools/core";
import { ParserRule } from "../types.js";
import { AttributeVisitorMixin } from "./rule-utils.js";
class AttributeValuesRequireQuotesVisitor extends AttributeVisitorMixin {
    checkStaticAttributeStaticValue({ attributeName, attributeValue, attributeNode }) {
        if (this.hasAttributeValue(attributeNode))
            return;
        if (this.isQuoted(attributeNode))
            return;
        this.addOffense(`Attribute value should be quoted: \`${attributeName}="${attributeValue}"\`. Always wrap attribute values in quotes.`, attributeNode.value.location, {
            node: attributeNode,
            unquotedValue: attributeValue
        });
    }
    checkStaticAttributeDynamicValue({ attributeName, attributeNode, combinedValue }) {
        if (this.hasAttributeValue(attributeNode))
            return;
        if (this.isQuoted(attributeNode))
            return;
        this.addOffense(`Attribute value should be quoted: \`${attributeName}="${combinedValue}"\`. Always wrap attribute values in quotes.`, attributeNode.value.location, {
            node: attributeNode,
            unquotedValue: combinedValue || ""
        });
    }
    hasAttributeValue(attributeNode) {
        return attributeNode.value?.type !== "AST_HTML_ATTRIBUTE_VALUE_NODE";
    }
    isQuoted(attributeNode) {
        const valueNode = attributeNode.value;
        return valueNode ? valueNode.quoted : false;
    }
}
export class HTMLAttributeValuesRequireQuotesRule extends ParserRule {
    static autocorrectable = true;
    name = "html-attribute-values-require-quotes";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error"
        };
    }
    check(result, context) {
        const visitor = new AttributeValuesRequireQuotesVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
    autofix(offense, result, _context) {
        if (!offense.autofixContext)
            return null;
        const { node: { value } } = offense.autofixContext;
        if (!value)
            return null;
        const quote = Token.from({ type: "TOKEN_QUOTE", value: '"', location: Location.zero, range: [0, 0] });
        if (value.open_quote) {
            value.open_quote.value = '"';
        }
        else {
            value.open_quote = quote;
        }
        if (value.close_quote) {
            value.close_quote.value = '"';
        }
        else {
            value.close_quote = quote;
        }
        value.quoted = true;
        return result;
    }
}
