// @herb-tools/linter/dist/src/rules/html-attribute-equals-spacing.js
import { BaseRuleVisitor } from "./rule-utils.js";
import { ParserRule } from "../types.js";
class HTMLAttributeEqualsSpacingVisitor extends BaseRuleVisitor {
    visitHTMLAttributeNode(attribute) {
        if (!attribute.equals || !attribute.name || !attribute.value) {
            return;
        }
        if (attribute.equals.value.startsWith(" ")) {
            this.addOffense("Remove whitespace before `=` in HTML attribute", attribute.equals.location, { node: attribute });
        }
        if (attribute.equals.value.endsWith(" ")) {
            this.addOffense("Remove whitespace after `=` in HTML attribute", attribute.equals.location, { node: attribute });
        }
    }
}
export class HTMLAttributeEqualsSpacingRule extends ParserRule {
    static autocorrectable = true;
    name = "html-attribute-equals-spacing";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error"
        };
    }
    check(result, context) {
        const visitor = new HTMLAttributeEqualsSpacingVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
    autofix(offense, result, _context) {
        if (!offense.autofixContext)
            return null;
        const { node: { equals } } = offense.autofixContext;
        if (!equals)
            return null;
        equals.value = "=";
        return result;
    }
}
