// @herb-tools/linter/dist/src/rules/html-attribute-double-quotes.js
import { ParserRule } from "../types.js";
import { AttributeVisitorMixin, getAttributeValueQuoteType, hasAttributeValue } from "./rule-utils.js";
import { filterLiteralNodes } from "@herb-tools/core";
class AttributeDoubleQuotesVisitor extends AttributeVisitorMixin {
    checkStaticAttributeStaticValue({ attributeName, attributeValue, attributeNode }) {
        if (!hasAttributeValue(attributeNode))
            return;
        if (getAttributeValueQuoteType(attributeNode) !== "single")
            return;
        if (attributeValue?.includes('"'))
            return;
        this.addOffense(`Attribute \`${attributeName}\` uses single quotes. Prefer double quotes for HTML attribute values: \`${attributeName}="${attributeValue}"\`.`, attributeNode.value.location, {
            node: attributeNode,
            valueContent: attributeValue
        });
    }
    checkStaticAttributeDynamicValue({ attributeName, valueNodes, attributeNode, combinedValue }) {
        if (!hasAttributeValue(attributeNode))
            return;
        if (getAttributeValueQuoteType(attributeNode) !== "single")
            return;
        if (filterLiteralNodes(valueNodes).some(node => node.content?.includes('"')))
            return;
        this.addOffense(`Attribute \`${attributeName}\` uses single quotes. Prefer double quotes for HTML attribute values: \`${attributeName}="${combinedValue}"\`.`, attributeNode.value.location, {
            node: attributeNode,
            valueContent: combinedValue || ""
        });
    }
}
export class HTMLAttributeDoubleQuotesRule extends ParserRule {
    static autocorrectable = true;
    name = "html-attribute-double-quotes";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "warning"
        };
    }
    check(result, context) {
        const visitor = new AttributeDoubleQuotesVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
    autofix(offense, result, _context) {
        if (!offense.autofixContext)
            return null;
        const { node: { value } } = offense.autofixContext;
        if (!value)
            return null;
        if (!value.open_quote)
            return null;
        if (!value.close_quote)
            return null;
        value.open_quote.value = '"';
        value.close_quote.value = '"';
        return result;
    }
}
