// @herb-tools/linter/dist/src/rules/html-body-only-elements.js
import { ParserRule } from "../types.js";
import { BaseRuleVisitor, getTagName, isBodyOnlyTag } from "./rule-utils.js";
class HTMLBodyOnlyElementsVisitor extends BaseRuleVisitor {
    elementStack = [];
    visitHTMLElementNode(node) {
        const tagName = getTagName(node.open_tag)?.toLowerCase();
        if (!tagName)
            return;
        this.checkBodyOnlyElement(node, tagName);
        this.elementStack.push(tagName);
        this.visitChildNodes(node);
        this.elementStack.pop();
    }
    checkBodyOnlyElement(node, tagName) {
        if (this.insideBody)
            return;
        if (!this.insideHead)
            return;
        if (!isBodyOnlyTag(tagName))
            return;
        this.addOffense(`Element \`<${tagName}>\` must be placed inside the \`<body>\` tag.`, node.location);
    }
    get insideBody() {
        return this.elementStack.includes("body");
    }
    get insideHead() {
        return this.elementStack.includes("head");
    }
}
export class HTMLBodyOnlyElementsRule extends ParserRule {
    static autocorrectable = false;
    name = "html-body-only-elements";
    get defaultConfig() {
        return {
            enabled: true,
            severity: "error",
            exclude: ["**/*.xml", "**/*.xml.erb"]
        };
    }
    check(result, context) {
        const visitor = new HTMLBodyOnlyElementsVisitor(this.name, context);
        visitor.visit(result.value);
        return visitor.offenses;
    }
}
